classes/search/cms.php

```diff
@@ -0,0 +1,188 @@
+            debugging('Error retrieving ' . $this->areaid . ' ' . $record->id . ' document, not all required data is available: ' .
+                $ex->getMessage(), DEBUG_DEVELOPER);
+            return false;
+        } catch (\dml_exception $ex) {
+            // Notify it as we run here as admin, we should see everything.
+            debugging('Error retrieving ' . $this->areaid . ' ' . $record->id . ' document: ' . $ex->getMessage(), DEBUG_DEVELOPER);
+            return false;
+        }
+
+        // Prepare associative array with data from DB.
+        $doc = \core_search\document_factory::instance($record->id, $this->componentname, $this->areaname);
+        $doc->set('title', content_to_text($record->value, false));
+        $doc->set('content', content_to_text($record->value, $record->valueformat));
+        $doc->set('contextid', $context->id);
+        $doc->set('courseid', $record->courseid);
+        $doc->set('owneruserid', \core_search\manager::NO_OWNER_ID);
+        $doc->set('modified', $record->timemodified);
+
+        // Check if this document should be considered new.
+        if (isset($options['lastindexedtime']) && ($options['lastindexedtime'] < $record->timecreated)) {
+            // If the document was created after the last index time, it must be new.
+            $doc->set_is_new(true);
+        }
+
+        return $doc;
+    }
+
+    /**
+     * Whether the user can access the document or not.
+     *
+     * @param int $id data id
+     * @return bool
+     */
+    public function check_access($id) {
+        try {
+            $data = $this->get_data($id);
+            $cminfo = $this->get_cm('cms', $data->instanceid, $data->courseid);
+        } catch (\dml_missing_record_exception $ex) {
+            return \core_search\manager::ACCESS_DELETED;
+        } catch (\dml_exception $ex) {
+            return \core_search\manager::ACCESS_DENIED;
+        }
+
+        // Recheck uservisible although it should have already been checked in core_search.
+        if ($cminfo->uservisible === false) {
+            return \core_search\manager::ACCESS_DENIED;
+        }
+
+        $context = \context_module::instance($cminfo->id);
+
+        if (!has_capability('mod/cms:view', $context)) {
+            return \core_search\manager::ACCESS_DENIED;
+        }
+
+        return \core_search\manager::ACCESS_GRANTED;
+    }
+
+    /**
+     * Link to the cms.
+     *
+     * @param \core_search\document $doc
+     * @return \moodle_url
+     */
+    public function get_doc_url(\core_search\document $doc) {
+        return $this->get_context_url($doc);
+    }
+
+    /**
+     * Link to the cms.
+     *
+     * @param \core_search\document $doc
+     * @return \moodle_url
+     */
+    public function get_context_url(\core_search\document $doc) {
+        $contextmodule = \context::instance_by_id($doc->get('contextid'));
+        return new \moodle_url('/mod/cms/view.php', ['id' => $contextmodule->instanceid]);
```

Comment:
I would suggest to double what the client expects to see as a link to found documents in the search results.

Reply:
This is just link and actual result is in doc https://github.com/catalyst/moodle-mod_cms/pull/129/files#diff-965e63736ed197b659eeb7349499abdfc8e2aa372cf1fd0a8ddb6b3467f1bf5aR98

Reply:
Yes and what about the link in the search result? It should link to some page, shouldn't it?

Reply:
I'm asking to the client what they expect.
Meanwhile, I'll leave same here and update other parts of this PR.